diff --git a/src/include/ifaddrs.c b/src/include/ifaddrs.c
index 109f322..8f30d4d 100644
--- a/src/include/ifaddrs.c
+++ b/src/include/ifaddrs.c
@@ -25,7 +25,9 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include "ifaddrs.h"
 
 #include <string.h>
+#define _GNU_SOURCE
 #include <stdlib.h>
+#include <stddef.h>
 #include <errno.h>
 #include <unistd.h>
 #include <sys/socket.h>
@@ -33,6 +35,7 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include <netinet/in.h>
 #include <linux/netlink.h>
 #include <linux/rtnetlink.h>
+#include <linux/if_packet.h>
 
 typedef struct NetlinkList
 {
diff --git a/src/include/lxcmntent.c b/src/include/lxcmntent.c
index dcb34a2..3b1a15b 100644
--- a/src/include/lxcmntent.c
+++ b/src/include/lxcmntent.c
@@ -18,6 +18,8 @@
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
  */
 
+#define _GNU_SOURCE
+#include <stdlib.h>
 #include <stdio.h>
 #include <string.h>
 #include <mntent.h>
diff --git a/src/include/openpty.c b/src/include/openpty.c
index 01579c5..3dc9bce 100644
--- a/src/include/openpty.c
+++ b/src/include/openpty.c
@@ -21,11 +21,11 @@
  *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
  */
 
-#define _XOPEN_SOURCE       /* See feature_test_macros(7) */
+#define _GNU_SOURCE
+#include <stdlib.h>
 #include <errno.h>
 #include <fcntl.h>
 #include <limits.h>
-#include <stdlib.h>
 #include <string.h>
 #include <termios.h>
 #include <unistd.h>
diff --git a/src/lxc/Makefile.am b/src/lxc/Makefile.am
index 92841aa..ebb0bf7 100644
--- a/src/lxc/Makefile.am
+++ b/src/lxc/Makefile.am
@@ -23,12 +23,12 @@ noinst_HEADERS = \
 	state.h \
 	utils.h
 
-if IS_BIONIC
+#if IS_BIONIC
 noinst_HEADERS += \
 	../include/ifaddrs.h \
 	../include/openpty.h \
 	../include/lxcmntent.h
-endif
+#endif
 
 if !HAVE_GETLINE
 if HAVE_FGETLN
@@ -97,12 +97,12 @@ if ENABLE_CGMANAGER
 liblxc_so_SOURCES += cgmanager.c
 endif
 
-if IS_BIONIC
+#if IS_BIONIC
 liblxc_so_SOURCES += \
 	../include/ifaddrs.c ../include/ifaddrs.h \
 	../include/openpty.c ../include/openpty.h \
 	../include/lxcmntent.c ../include/lxcmntent.h
-endif
+#endif
 
 if !HAVE_GETLINE
 if HAVE_FGETLN
diff --git a/src/lxc/lxccontainer.c b/src/lxc/lxccontainer.c
old mode 100644
new mode 100755
index 2582da0..34f6370
--- a/src/lxc/lxccontainer.c
+++ b/src/lxc/lxccontainer.c
@@ -18,6 +18,13 @@
  *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
  */
 
+#if HAVE_IFADDRS_H
+#include <ifaddrs.h>
+#else
+#include <../include/ifaddrs.h>
+#endif
+
+
 #define _GNU_SOURCE
 #include <assert.h>
 #include <stdarg.h>
@@ -40,6 +47,8 @@
 #include <lxc/lxccontainer.h>
 #include <lxc/version.h>
 
+#include <../include/getline.h>
+
 #include "config.h"
 #include "lxc.h"
 #include "state.h"
@@ -56,12 +65,6 @@
 #include "namespace.h"
 #include "lxclock.h"
 
-#if HAVE_IFADDRS_H
-#include <ifaddrs.h>
-#else
-#include <../include/ifaddrs.h>
-#endif
-
 #define MAX_BUFFER 4096
 
 #define NOT_SUPPORTED_ERROR "the requested function %s is not currently supported with unprivileged containers"
@@ -79,6 +82,15 @@ return -1;
 }
 #endif
 
+static inline int setns(int fd, int nstype)
+{
+#ifdef __NR_setns
+return syscall(__NR_setns, fd, nstype);
+#else
+errno = ENOSYS;
+return -1;
+#endif
+}
 
 lxc_log_define(lxc_container, lxc);
 
diff --git a/src/lxc/network.c b/src/lxc/network.c
index a9900de..8e6dbb1 100644
--- a/src/lxc/network.c
+++ b/src/lxc/network.c
@@ -20,6 +20,13 @@
  * License along with this library; if not, write to the Free Software
  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
  */
+
+#if HAVE_IFADDRS_H
+#include <ifaddrs.h>
+#else
+#include <../include/ifaddrs.h>
+#endif
+
 #define _GNU_SOURCE
 #include <stdio.h>
 #undef _GNU_SOURCe
@@ -49,12 +56,6 @@
 #include "network.h"
 #include "conf.h"
 
-#if HAVE_IFADDRS_H
-#include <ifaddrs.h>
-#else
-#include <../include/ifaddrs.h>
-#endif
-
 #ifndef IFLA_LINKMODE
 #  define IFLA_LINKMODE 17
 #endif
